rhp/v4/rhp.go

```diff
@@ -0,0 +1,754 @@
+	return hp.EgressPrice.Mul64(SectorSize)
+}
+
+// RPCModifySectorsCost returns the cost of modifying a contract's sectors with the
+// given actions. The duration parameter is the number of blocks until the
+// contract's expiration height.
+func (hp HostPrices) RPCModifySectorsCost(actions []ModifyAction) (cost types.Currency) {
+	var n uint64
+	for _, action := range actions {
+		if action.Type == ActionTrim {
+			n += action.N
+		} else {
+			n++
+		}
+	}
+	return hp.ModifySectorActionPrice.Mul64(n)
+}
+
+// RPCAppendSectorsCost returns the cost of appending sectors to a contract. The duration
+// parameter is the number of blocks until the contract's expiration height.
+func (hp HostPrices) RPCAppendSectorsCost(sectors, duration uint64) (cost, collateral types.Currency) {
+	storage, collateral := hp.StoreSectorCost(duration)
+	storage = storage.Mul64(sectors)
+	collateral = collateral.Mul64(sectors)
+	return storage, collateral
+}
+
+// SigHash returns the hash of the host settings used for signing.
+func (hp HostPrices) SigHash() types.Hash256 {
+	h := types.NewHasher()
+	types.V2Currency(hp.ContractPrice).EncodeTo(h.E)
+	types.V2Currency(hp.Collateral).EncodeTo(h.E)
+	types.V2Currency(hp.StoragePrice).EncodeTo(h.E)
+	types.V2Currency(hp.IngressPrice).EncodeTo(h.E)
+	types.V2Currency(hp.EgressPrice).EncodeTo(h.E)
+	types.V2Currency(hp.ModifySectorActionPrice).EncodeTo(h.E)
+	h.E.WriteUint64(hp.TipHeight)
+	h.E.WriteTime(hp.ValidUntil)
+	return h.Sum()
+}
+
+// Validate checks the host prices for validity. It returns an error if the
+// prices have expired or the signature is invalid.
+func (hp *HostPrices) Validate(pk types.PublicKey) error {
+	if time.Until(hp.ValidUntil) <= 0 {
+		return ErrPricesExpired
+	}
+	if !pk.VerifyHash(hp.SigHash(), hp.Signature) {
+		return ErrInvalidSignature
+	}
+	return nil
+}
+
+// HostSettings specify the settings of a host.
+type HostSettings struct {
+	ProtocolVersion     [3]uint8       `json:"protocolVersion"`
+	Release             string         `json:"release"`
+	WalletAddress       types.Address  `json:"walletAddress"`
+	AcceptingContracts  bool           `json:"acceptingContracts"`
+	MaxCollateral       types.Currency `json:"maxCollateral"`
+	MaxContractDuration uint64         `json:"maxContractDuration"`
+	MaxSectorDuration   uint64         `json:"maxSectorDuration"`
+	MaxModifyActions    uint64         `json:"maxModifyActions"`
+	RemainingStorage    uint64         `json:"remainingStorage"`
+	TotalStorage        uint64         `json:"totalStorage"`
+	Prices              HostPrices     `json:"prices"`
+}
+
+// A ModifyAction adds or modifies sectors within a contract.
+type ModifyAction struct {
+	Type uint8         `json:"type"`
+	Root types.Hash256 `json:"root,omitempty"` // Update
+	A    uint64        `json:"a,omitempty"`    // Swap
+	B    uint64        `json:"b,omitempty"`    // Swap
+	N    uint64        `json:"n,omitempty"`    // Trim, Update
+}
```

Comment:
How does `ActionUpdate` work? From the server code in `coreutils`, it looks like it overwrites the sector root at index `A` with `Root`. That's very different from the old `RPCWriteActionUpdate`, which overwrote sector *data*. Also, if `Root` is the root of a sector that the host isn't storing, the host won't be able to complete a storage proof for it!

I think we should consider scrapping the general-purpose `ModifyAction` design and instead hard-code the specific actions we know we want. When RHPv2 was designed, we envisioned something more like a distributed filesystem: sector 0 was used to store metadata about the other sectors, and diff proofs allowed you to atomically update both the data and the metadata. Nowadays (AFAICT) we only do two things: append new sectors, and delete old sectors (via swap + trim). Append is already a separate RPC now; perhaps instead of a generic `RPCModify`, we only need `RPCSwapAndTrim`. That would simplify the diff proofs a bit too.

More radically, we could get rid of trimming. Instead, "deleted" sectors are overwritten with zeroes, and when you upload new sectors, they replace zeroed sectors if available. This matches how the payment works anyway: renters don't get refunded when they trim. I think we discussed this option at the RHP summit, and presumably we rejected it, but I don't remember the specific objections.

Reply:
The presumption is that the sector will always exist on the host already through RPC write. Similar to append, the server should be checking if the host has a sector before allowing the update.

Update is currently necessary for cost efficient deletion. After trimming, there isn’t a way to track the max size of the contract so future appends will cost additional SC and we can’t do the atomic append+swap+trim any more. It does feel out of place in the new rpc design.